### PR TITLE
Auth-core module

### PR DESCRIPTION
**Pull request by StuartJ45**
PR for issue #59. 
Created an auth-core module and extracted all common / non-akka functionality into it.

### CODE REVIEW

File: keycloak4s-auth/core/src/main/scala/com/fullfacing/keycloak4s/auth/core/authorization/Authorization.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  val service: String
+
+  /**
+   * Abstract function to handle authorization of a request.
+   *
+   * @param requestPath The path of the HTTP request.
+   * @param method      The HTTP method of the request.
+   * @param userRoles   The permissions of the user.
+   */
+  def authorizeRequest(requestPath: String, method: String, userRoles: List[String])(implicit cId: UUID): Boolean
```
Comment (Executioner1939) [inline keycloak4s-auth/core/src/main/scala/com/fullfacing/keycloak4s/auth/core/authorization/Authorization.scala]: Maybe make this generic, allowing the implementer to specify the Request type, which allows access to the full scope of the request.

**Comment by Fristi** [comment]
For this JWS decoding, is it possible to use a scala library like https://github.com/pauldijou/jwt-scala ?

**Comment by Executioner1939** [comment]
> For this JWS decoding, is it possible to use a scala library like https://github.com/pauldijou/jwt-scala ?

I am for moving to this library, if everyone else is in agreement? @lmuller90 @StuartJ45 @neil-fladderak

**Comment by StuartJ45** [comment]
> > For this JWS decoding, is it possible to use a scala library like https://github.com/pauldijou/jwt-scala ?
> 
> I am for moving to this library, if everyone else is in agreement? @lmuller90 @StuartJ45 @neil-fladderak

@lmuller90 will have to confirm on this. I believe he used the nimbus library's built in function for retrieving the public key to verify the signature. Although I'm sure that functionality can be recreated if we go for the scala library.

**Comment by Executioner1939** [comment]
> > > For this JWS decoding, is it possible to use a scala library like https://github.com/pauldijou/jwt-scala ?
> > 
> > 
> > I am for moving to this library, if everyone else is in agreement? @lmuller90 @StuartJ45 @neil-fladderak
> 
> @lmuller90 will have to confirm on this. I believe he used the nimbus library's built in function for retrieving the public key to verify the signature. Although I'm sure that functionality can be recreated if we go for the scala library.

This is correct, I don't see any native way of handling JWK's and JWKS's in the proposed library. One of the reasons for using Nimbus was that it is quite a thorough implementation of the full JOSE and JWT standards.

**Comment by lmuller90** [comment]
> > > > For this JWS decoding, is it possible to use a scala library like [pauldijou/jwt-scala](https://github.com/pauldijou/jwt-scala) ?
> > > 
> > > 
> > > I am for moving to this library, if everyone else is in agreement? @lmuller90 @StuartJ45 @neil-fladderak
> > 
> > 
> > @lmuller90 will have to confirm on this. I believe he used the nimbus library's built in function for retrieving the public key to verify the signature. Although I'm sure that functionality can be recreated if we go for the scala library.
> 
> This is correct, I don't see any native way of handling JWK's and JWKS's in the proposed library. One of the reasons for using Nimbus was that it is quite a thorough implementation of the full JOSE and JWT standards.

I'll have to take a good look at all the Scala JWT libraries available again, last time I played around with them I decided to stick with Nimbus because I wanted to avoid delaying our initial release.

I propose we leave that for a separate PR, as long as the interface for the token validator remains more or less the same it shouldn't affect the code for the client adapters much.

File: keycloak4s-auth/core/src/main/scala/com/fullfacing/keycloak4s/auth/core/authorization/Authorization.scala (hunk 1 above)
Comment (StuartJ45) [reply keycloak4s-auth/core/src/main/scala/com/fullfacing/keycloak4s/auth/core/authorization/Authorization.scala]: Updated

**Pull request merged**